trestle/core/control_interface.py

```diff
@@ -1109,13 +1109,23 @@ def insert_imp_req_into_component(
                     if set_param.param_id not in (control_imp_rule_param_ids + imp_req_rule_param_ids):
                         continue
                     found = False
-                    for dest_param in imp_req.set_parameters:
+                    for dest_param in as_list(imp_req.set_parameters):
                         if dest_param.param_id != set_param.param_id:
                             continue
                         dest_param.values = set_param.values
                         found = True
                         break
                     # if rule parameter val was not already set by a set_param, make new set_param for it
+                    if found:
+                        continue
+                    # but first check if the parameter was already set with the same value in the control_imp
+                    # if so we don't need to insert a new set_param in imp_req
+                    for dest_param in as_list(control_imp.set_parameters):
+                        if dest_param.param_id != set_param.param_id:
+                            continue
+                        if dest_param.values == set_param.values:
+                            found = True
+                            break
                     if found:
```

Comment:
Hmm the `found` flag is set to true in the for loop over implementation requirement's set_parameters. Do we want to reset it before we ran the second loop?

Reply:
The logic is ok because it only falls through to the next loop if found is still false.  There are two places a setparam might hide and only if it isn't in both do you fall through and make a new one.  And even if there is a setparam at the high level - if its value is not the needed value then you need to make a new setparam.